R/CreatePackageReport.R

```diff
@@ -1,3 +1,261 @@
+                                    pattern = pkg_name
+                                    , fileext = ".html"
+                                  )
+                              ) {
+            # Input validation for pkg_name, pkg_path
+            # report_path validated by its active binding
+            assertthat::assert_that(
+                assertthat::is.string(pkg_name)
+                , pkg_name != ""
+                , is.null(pkg_path) || assertthat::is.readable(pkg_path)
+            )
+            .validate_pkg_name(pkg_name)
+
+            private$protected$pkg_name <- pkg_name
+            private$protected$pkg_path <- pkg_path
+            self$report_path <- report_path
+
+            return(invisible(self))
+        }
+
+        , render_report = function() {
+            log_info("Rendering package report...")
+
+            silence_logger()
+            rmarkdown::render(
+                system.file(
+                    file.path("package_report", "package_report.Rmd")
+                    , package = "pkgnet"
+                )
+                , output_dir = dirname(self$report_path)
+                , output_file = basename(self$report_path)
+                , quiet = TRUE
+                , params = list(
+                    reporters = private$reporters
+                    , pkg_name = self$pkg_name
+                )
+            )
+            unsilence_logger()
+
+            log_info(paste(
+                "Done creating package report!"
+                , sprintf("It is available at %s", self$report_path)
+            ))
+
+            # If suppress flag is unset, then env variable will be emptry string ""
+            if (identical(Sys.getenv("PKGNET_SUPPRESS_BROWSER"), "")) {
+                utils::browseURL(self$report_path)
+            }
+
+            return(invisible(self))
+        }
+
+    ) # / public
+
+    , active = list(
+        pkg_name = function() {
+            return(private$protected$pkg_name)
+        }
+        , pkg_path = function() {
+            return(private$protected$pkg_path)
+        }
+        , report_path = function(report_path) {
+            if (!missing(report_path)) {
+                assertthat::assert_that(
+                    assertthat::is.string(report_path)
+                    , report_path != ""
+                )
+                if (!identical(tolower(tools::file_ext(report_path)), "html")){
+                    log_fatal(paste(
+                        "report_path must be a .html file path."
+                        , sprintf("You gave '%s'.", report_path)
+                    ))
+                }
+                assertthat::assert_that(
+                    assertthat::is.writeable(dirname(report_path))
+                )
+
+                private$protected$report_path <- report_path
+            }
+            return(private$protected$report_path)
+        }
+        , SummaryReporter = function(reporter) {
```

Comment:
ugh this is one of those cases where GitHub's PR review is BRUTAL compared to Bitbucket. It's hard for me to see which files got touched.

Here's my question: why was it necessary to have named active bindings for each of these? Since we're just `lapply`-ing over "whatever reporters you give us" (https://github.com/UptakeOpenSource/pkgnet/blob/master/inst/package_report/package_report.Rmd#L31), and you aren't actually doing any special handling here, why is it necessary to have `DependencyReporter`, `FunctionReporter`, etc.

Reply:
So there is special handling, but I ended up moving it all into `private$set_reporter` so I could repeat less code. If we do this with public fields instead, then there's no way to prevent someone from adding a non-reporter object using `<-` assignment right?